azure/services/natgateways/natgateways.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package natgateways
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-06-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha4"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+)
+
+// NatGatewayScope defines the scope interface for nat gateway service
+type NatGatewayScope interface {
+	logr.Logger
+	azure.ClusterScoper
+	NatGatewaySpecs() []azure.NatGatewaySpec
+	SetNodeNatGateway(natGateway infrav1.NatGateway)
+}
+
+// Service provides operations on azure resources
+type Service struct {
+	Scope NatGatewayScope
+	client
+}
+
+// New creates a new service.
+func New(scope NatGatewayScope) *Service {
+	return &Service{
+		Scope:  scope,
+		client: newClient(scope),
+	}
+}
+
+// Reconcile gets/creates/updates a nat gateway.
+// Only when the Nat Gateway 'Name' property is defined we create the Nat Gateway: it's opt-in.
+func (s *Service) Reconcile(ctx context.Context) error {
+	ctx, span := tele.Tracer().Start(ctx, "natgateways.Service.Reconcile")
+	defer span.End()
+
+	if !s.Scope.Vnet().IsManaged(s.Scope.ClusterName()) {
+		s.Scope.V(4).Info("Skipping nat gateways reconcile in custom vnet mode")
+		return nil
+	}
+
+	for _, natGatewaySpec := range s.Scope.NatGatewaySpecs() {
+		existingNatGateway, err := s.getExisting(ctx, natGatewaySpec)
+
+		switch {
+		case err != nil && !azure.ResourceNotFound(err):
+			return errors.Wrapf(err, "failed to get nat gateway %s in %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
+		case err == nil:
+			// nat gateway already exists
+			s.Scope.V(4).Info("nat gateway already exists", "nat gateway", natGatewaySpec.Name)
+			s.Scope.SetNodeNatGateway(*existingNatGateway)
+
+			var update bool
+			if existingNatGateway.NatGatewayIP.Name != natGatewaySpec.NatGatewayIP.Name {
+				update = true
```

Comment:
I don't see update being used elsewhere in the code block. Perhaps, move the `!update` block into here and invert the if statement.

Reply:
Done.